components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs

```diff
@@ -440,32 +556,66 @@ public class LegacyJobOrchestrator : IJobOrchestrator, IPublisherConfigServices,
             var sw = Stopwatch.StartNew();
             await _lockConfig.WaitAsync(ct).ConfigureAwait(false);
             try {
-                var nodeFound = false;
-                var existingGroup = new List<PublishedNodesEntryModel>();
+                // ToDo: Uncomment ValidateRequest() once our test requests pass validation.
+                // This will throw SerializerException if values of request fields are not conformant.
+                //ValidateRequest(request);
+
+                var dataSetFound = false;
+                var existingGroups = new List<PublishedNodesEntryModel>();
                 foreach (var entry in _publishedNodesEntrys) {
                     if (entry.HasSameGroup(request)) {
-                        if (request.DataSetWriterId == entry.DataSetWriterId &&
-                            request.DataSetPublishingInterval == entry.DataSetPublishingInterval) {
-                            entry.OpcNodes.AddRange(request.OpcNodes);
-                            nodeFound = true;
+                        // We may have several entries with the same DataSetGroup definition,
+                        // so we will add nodes only if the whole DataSet definition matches.
+                        if (IsSameDataSet(entry, request)) {
+                            // Create HashSet of nodes for this entry.
+                            var existingNodesSet = new HashSet<OpcNodeModel>(OpcNodeModelEx.Comparer);
+                            existingNodesSet.UnionWith(entry.OpcNodes);
+
+                            foreach (var nodeToAdd in request.OpcNodes) {
+                                if (!existingNodesSet.Contains(nodeToAdd)) {
+                                    entry.OpcNodes.Add(nodeToAdd);
+                                    existingNodesSet.Add(nodeToAdd);
+                                }
+                                else {
+                                    _logger.Debug("Node \"{node}\" is already present for entry with \"{endpoint}\" endpoint.",
+                                        nodeToAdd.Id, entry.EndpointUrl);
+                                }
+                            }
+
+                            dataSetFound = true;
                         }
-                        existingGroup.Add(entry);
+
+                        // Even if DataSets did not match, we need to add this entry to existingGroups
+                        // so that generated job definition is complete.
+                        existingGroups.Add(entry);
                     }
                 }
 
-                if (!nodeFound) {
-                    existingGroup.Add(request);
+                if (!dataSetFound) {
+                    existingGroups.Add(request);
                     _publishedNodesEntrys.Add(request);
                 }
-                
+
+                // Update published nodes file.
+                var updatedContent = JsonSerializer.Serialize(_publishedNodesEntrys);
```

Comment:
Let's have a separate function that persists `_publishedNodesEntrys` to published nodes file.

Reply:
Added.